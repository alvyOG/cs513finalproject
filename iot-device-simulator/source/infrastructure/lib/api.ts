
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import {
  AccessLogFormat,
  AuthorizationType,
  ContentHandling,
  Deployment,
  EndpointType,
  Integration,
  IntegrationType,
  LogGroupLogDestination,
  MethodLoggingLevel,
  MethodOptions,
  PassthroughBehavior,
  RequestValidator,
  RestApi,
  Stage
} from '@aws-cdk/aws-apigateway';
import { Policy, ServicePrincipal } from '@aws-cdk/aws-iam';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { addCfnSuppressRules } from '../utils/utils';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';
import { IBucket } from '@aws-cdk/aws-s3';
import { Table } from '@aws-cdk/aws-dynamodb';

/**
 * ApiConstructProps props
 * @interface ApiConstructProps
 */
export interface ApiConstructProps {
  // Policy for CloudWatch Logs
  readonly cloudWatchLogsPolicy: Policy;
  // IoT endpoint address
  readonly stepFunctionsARN: string;
  // Simulation data DynamoDB table name
  readonly simulationTable: Table;
  // Device Type data DynamoDB table name
  readonly deviceTypeTable: Table;
  //Routes S3 bucket
  readonly routesBucketArn: string;
  //microservices lambda
  readonly microservicesLambda: LambdaFunction;
  /**
   * Solution config properties.
   * Logging level, solution ID, version, source code bucket, and source code prefix
   */
  readonly solutionConfig: {
    sendAnonymousUsage: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
  };
  // Solution UUID
  readonly uuid: string;
}

/**
 * @class
 * IoT Device Simulator Framework API Construct.
 * It creates an API Gateway REST API and other resources.
 */
export class ApiConstruct extends Construct {
  // API endpoint
  public apiEndpoint: string;
  // API ID
  public apiId: string;
  //microservices lambda
  public microservicesLambdaFunction: LambdaFunction;

  constructor(scope: Construct, id: string, props: ApiConstructProps) {
    super(scope, id);

    const apiLogGroup = new LogGroup(this, 'Logs', {
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.THREE_MONTHS
    });
    addCfnSuppressRules(apiLogGroup, [{
      id: 'W84', reason: 'CloudWatch Logs are already encrypted by default.'
    }]);

    const api = new RestApi(this, 'IoTDeviceSimulatorApi', {
      defaultCorsPreflightOptions: {
        allowOrigins: ['*'],
        allowHeaders: [
          'Authorization',
          'Content-Type',
          'X-Amz-Date',
          'X-Amz-Security-Token',
          'X-Api-Key'
        ],
        allowMethods: [
          'GET',
          'POST',
          'PUT',
          'DELETE',
          'OPTIONS'
        ],
        statusCode: 200
      },
      deploy: true,
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(apiLogGroup),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'prod',
        tracingEnabled: true
      },
      description: 'IoT Device Simulator Rest API',
      endpointTypes: [EndpointType.REGIONAL]
    });
    this.apiEndpoint = `https://${api.restApiId}.execute-api.${Aws.REGION}.amazonaws.com/prod`;
    this.apiId = api.restApiId;

    const requestValidator = new RequestValidator(this, 'ApiRequestValidator', {
      restApi: api,
      validateRequestParameters: true,
      validateRequestBody: true
    });

    addCfnSuppressRules(api.node.findChild('Deployment') as Deployment, [{
      id: 'W68', reason: 'The solution does not require the usage plan.'
    }]);
    addCfnSuppressRules(api.node.findChild('DeploymentStage.prod') as Stage, [{
      id: 'W64', reason: 'The solution does not require the usage plan.'
    }]);

    /**
     * method options for all methods
     */
    const universalMethodOptions: MethodOptions = {
      authorizationType: AuthorizationType.IAM,
      methodResponses: [{
        statusCode: '200',
        responseModels: {
          'application/json': { modelId: 'Empty' }
        }
      }],
      requestParameters: { 'method.request.querystring.nextToken': false },
      requestValidator: requestValidator
    };

    /**
     * Integration for all resources
     */
    const universalIntegration = new Integration({
      type: IntegrationType.AWS_PROXY,
      integrationHttpMethod: 'POST',
      options: {
        contentHandling: ContentHandling.CONVERT_TO_TEXT,
        integrationResponses: [{ statusCode: '200' }],
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH
      },
      uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${props.microservicesLambda.functionArn}/invocations`
    });

    /**
     * simulation API
     * ANY /simulation
     * ANY /simulation/{simid}
     */
    const simulationResource = api.root.addResource('simulation');
    simulationResource.addMethod(
      'ANY',
      universalIntegration,
      universalMethodOptions
    );
    const simulationSimIdResource = simulationResource.addResource('{simid}');
    simulationSimIdResource.addMethod('ANY', universalIntegration, universalMethodOptions);

    /**
     * Devive Types API
     * ANY /devicetypes
     * ANY /devicetypes/{typeid}
     */
    const deviceTypesResource = api.root.addResource('devicetypes');
    deviceTypesResource.addMethod('ANY', universalIntegration, universalMethodOptions);
    const typeIdResource = deviceTypesResource.addResource('{typeid}');
    typeIdResource.addMethod('ANY', universalIntegration, universalMethodOptions);

    props.microservicesLambda.addPermission('ApiLambdaInvokePermission', {
      action: 'lambda:InvokeFunction',
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      sourceArn: api.arnForExecuteApi()
    });
  }
}