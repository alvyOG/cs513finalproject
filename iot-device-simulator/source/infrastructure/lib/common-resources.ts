// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, Construct, RemovalPolicy, Stack } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, IBucket } from '@aws-cdk/aws-s3';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * CommonResourcesConstruct props
 * @interface CommonResourcesConstructProps
 */
export interface CommonResourcesConstructProps {
  // Source code bucket
  sourceCodeBucket: string;
}

/**
 * @class
 * IoT Device Simulator Common Resources Construct.
 * It creates a common CloudWatch Logs policy for Lambda functions, a logging S3 bucket, and M2C2 S3 bucket.
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch Logs policy
  public cloudWatchLogsPolicy: Policy;
  // S3 Logging bucket
  public s3LoggingBucket: Bucket;
  // Code S3 bucket
  public sourceCodeBucket: IBucket;

  constructor(scope: Construct, id: string, props: CommonResourcesConstructProps) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new Policy(this, 'CloudWatchLogsPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: [
            Stack.of(this).formatArn({ service: 'logs', resource: 'log-group', resourceName: '/aws/lambda/*', arnFormat: ArnFormat.COLON_RESOURCE_NAME })
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });

    addCfnSuppressRules(this.s3LoggingBucket, [
      { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
      { id: 'W51', reason: 'This bucket is to store S3 logs, so it does not require S3 policy.' }
    ]);

    this.sourceCodeBucket = Bucket.fromBucketName(this, 'SourceCodeBucket', props.sourceCodeBucket);
  }
}