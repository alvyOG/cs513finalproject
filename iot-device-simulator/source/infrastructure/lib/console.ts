// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, ArnFormat, Construct, Duration, RemovalPolicy, Stack } from '@aws-cdk/core';
import { CfnIdentityPool, CfnIdentityPoolRoleAttachment, CfnUserPool, CfnUserPoolUser, UserPool, UserPoolClient } from '@aws-cdk/aws-cognito';
import { Effect, FederatedPrincipal, PolicyDocument, PolicyStatement, Role } from '@aws-cdk/aws-iam';
import { Bucket, IBucket } from '@aws-cdk/aws-s3';
import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';
import { CfnPolicy } from '@aws-cdk/aws-iot';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * UiConstructProps props
 * @interface ConsoleConstructProps
 */
export interface ConsoleConstructProps {
  // The API ID
  readonly apiId: string;
  // S3 logging bucket
  readonly s3LogsBucket: Bucket;
  // User E-Mail address
  readonly adminEmail: string;
  //map resources
  readonly mapArn: string;
  readonly placeIndexArn: string;
}

/**
 * @class
 * IoT Device Simulator Framework UI Construct.
 * It creates a CloudFront distribution, an UI hosting S3 bucket,
 * Cognito resources, and custom resources for UI assets.
 */
export class ConsoleConstruct extends Construct {
  // CloudFront distribution domain name
  public cloudFrontDomainName: string;
  // Cognito Identity pool ID
  public identityPoolId: string;
  // UI S3 bucket
  public consoleBucket: IBucket;
  // Cognito user pool ID
  public userPoolId: string;
  // Cognito user pool web client ID
  public webClientId: string;
  //IoT policy name
  public iotPolicy: CfnPolicy;

  constructor(scope: Construct, id: string, props: ConsoleConstructProps) {
    super(scope, id);

    const consoleCloudfrontDist = new CloudFrontToS3(this, 'Distribution', {
      bucketProps: {
        serverAccessLogsBucket: props.s3LogsBucket,
        serverAccessLogsPrefix: 'console-s3/'
      },
      cloudFrontDistributionProps: {
        comment: 'IoT Device Simulator Distribution',
        enableLogging: true,
        errorResponses: [
          { httpStatus: 403, responseHttpStatus: 200, responsePagePath: '/index.html' },
          { httpStatus: 404, responseHttpStatus: 200, responsePagePath: '/index.html' }
        ],
        httpVersion: 'http2',
        logBucket: props.s3LogsBucket,
        logFilePrefix: 'console-cf/'

      },
      insertHttpSecurityHeaders: false
    });
    this.cloudFrontDomainName = consoleCloudfrontDist.cloudFrontWebDistribution.domainName;
    this.consoleBucket = consoleCloudfrontDist.s3BucketInterface;

    const userPool = new UserPool(this, 'UserPool', {
      passwordPolicy: {
        minLength: 12,
        requireDigits: true,
        requireLowercase: true,
        requireSymbols: true,
        requireUppercase: true
      },
      removalPolicy: RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      signInAliases: {
        email: true
      },
      userPoolName: `${Aws.STACK_NAME}-user-pool`,
      userInvitation: {
        emailSubject: '[IoT Device Simulator] Login information',
        emailBody: `
          <p>
            You are invited to join IoT Device Simulator.<br />
            https://${this.cloudFrontDomainName}
          </p>
          <p>
            Please sign in to IoT Device Simulator using the temporary credentials below:<br />
            Username: <strong>{username}</strong><br />Password: <strong>{####}</strong>
          </p>
        `
      }
    });
    (userPool.node.defaultChild as CfnUserPool).userPoolAddOns = { advancedSecurityMode: 'ENFORCED' };
    this.userPoolId = userPool.userPoolId;

    const userPoolClient = new UserPoolClient(this, 'UserPoolClient', {
      generateSecret: false,
      preventUserExistenceErrors: true,
      refreshTokenValidity: Duration.days(1),
      userPool,
      userPoolClientName: `${Aws.STACK_NAME}-userpool-client`
    });
    this.webClientId = userPoolClient.userPoolClientId;

    new CfnUserPoolUser(this, 'User', { // NOSONAR: typescript:S1848
      userPoolId: userPool.userPoolId,
      desiredDeliveryMediums: ['EMAIL'],
      forceAliasCreation: true,
      userAttributes: [
        { name: 'email', value: props.adminEmail },
        { name: 'email_verified', value: 'true' }
      ],
      username: props.adminEmail
    });

    const identityPool = new CfnIdentityPool(this, 'IdentityPool', {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [{
        clientId: userPoolClient.userPoolClientId,
        providerName: userPool.userPoolProviderName,
        serverSideTokenCheck: false
      }]
    });


    this.identityPoolId = identityPool.ref;

    const idsIotPolicy = new CfnPolicy(this, 'IDS-IoT-Policy', {
      policyDocument: new PolicyDocument({
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Connect"],
            resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'client', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Subscribe"],
            resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topicfilter', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Receive"],
            resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topic', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
          })
        ]
      })
    });

    this.iotPolicy = idsIotPolicy;
    addCfnSuppressRules(idsIotPolicy, [
      { id: 'W11', reason: 'Cannot specify the resource to attach policy to identity.' }
    ]);

    const authenticatedRole = new Role(this, 'IdentityPoolAuthenticatedRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': { 'cognito-identity.amazonaws.com:amr': 'authenticated' }
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
      description: `${Aws.STACK_NAME} Identity Pool authenticated role`,
      inlinePolicies: {
        'ExecuteApiPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['execute-api:Invoke'],
              resources: [Stack.of(this).formatArn({ service: 'execute-api', resource: `${props.apiId}`, resourceName: 'prod/*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            }),
          ],
        }),
        'LocationServicePolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "geo:SearchPlaceIndexForText",
                "geo:GetMapGlyphs",
                "geo:GetMapSprites",
                "geo:GetMapStyleDescriptor",
                "geo:SearchPlaceIndexForPosition",
                "execute-api:Invoke",
                "geo:GetMapTile"
              ],
              resources: [
                props.mapArn,
                props.placeIndexArn,
              ]
            }),
          ]
        }),
        'IoTPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "iot:AttachPrincipalPolicy"
              ],
              resources: ['*']
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Connect"],
              resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'client', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Subscribe"],
              resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topicfilter', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Receive"],
              resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topic', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            })
          ]
        })
      }
    });
    addCfnSuppressRules(authenticatedRole, [
      { id: 'W11', reason: 'iot:AttachPrincipalPolicy does not allow for resource specification' }
    ]);

    new CfnIdentityPoolRoleAttachment(this, 'IdentityPoolRoleAttachement', { // NOSONAR: typescript:S1848
      identityPoolId: identityPool.ref,
      roles: { authenticated: authenticatedRole.roleArn }
    });
  }
}