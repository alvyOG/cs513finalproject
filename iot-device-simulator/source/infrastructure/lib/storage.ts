// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct, RemovalPolicy } from '@aws-cdk/core';
import { AttributeType, BillingMode, Table, TableEncryption } from '@aws-cdk/aws-dynamodb';
import { Bucket, BucketEncryption, BlockPublicAccess} from '@aws-cdk/aws-s3';
import { PolicyStatement, AnyPrincipal, Effect } from '@aws-cdk/aws-iam';

/**,
 * @interface StorageContructProps
 * StorageContruct props
 */
export interface StorageContructProps {
  readonly solutionId: string;
  readonly s3LogsBucket: Bucket
}

/**
 * IoT Device Simulator storage construct
 * Creates an S3 bucket to store test scenarios and 
 * a DynamoDB table to store tests and test configuration
 */
export class StorageContruct extends Construct {
  public simulationsTable: Table;
  public deviceTypesTable: Table;
  public routesBucket: Bucket

  constructor(scope: Construct, id: string, props: StorageContructProps) {
    super(scope, id);

    this.simulationsTable = new Table(this, 'IDS-Simulations-Table', {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: 'simId', type: AttributeType.STRING },
      pointInTimeRecovery: true
    });

    this.deviceTypesTable = new Table(this, 'IDS-Device-Types-Table', {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: 'typeId', type: AttributeType.STRING },
      pointInTimeRecovery: true
    });

    this.routesBucket = new Bucket(this, 'RoutesBucket', {
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: props.s3LogsBucket,
      serverAccessLogsPrefix: 'routes-bucket-access/',
      encryption: BucketEncryption.S3_MANAGED,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    this.routesBucket.addToResourcePolicy(
      new PolicyStatement({
        sid: 'HttpsOnly',
        resources: [
          `${this.routesBucket.bucketArn}/*`,
          `${this.routesBucket.bucketArn}`
        ],
        actions: ['*'],
        principals: [new AnyPrincipal()],
        effect: Effect.DENY,
        conditions:
        {
          Bool: {
            'aws:SecureTransport': 'false'
          }
        }
      })
    );
  }
}