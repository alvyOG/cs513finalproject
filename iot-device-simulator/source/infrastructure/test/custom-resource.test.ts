// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { CustomResourcesConstruct } from '../lib/custom-resource';


test('IoT Device Simulator customResourcesConstruct Test', () => {
    const stack = new Stack();
    const testRouteBucket = new Bucket(stack, 'testRouteBucket');
    const testConsoleBucket = new Bucket(stack, 'testConsoleBucket')
    const testPolicy = new Policy(stack, 'TestPolicy', {
        statements: [
            new PolicyStatement({
                resources: ['*'],
                actions: ['cloudwatch:Get*']
            })
        ]
    });
    const testSourceBucket = Bucket.fromBucketName(stack, 'SourceCodeBucket', 'test-bucket-region');

    const customResources = new CustomResourcesConstruct(stack, 'TestCustomResource', {
        cloudWatchLogsPolicy: testPolicy,
        solutionConfig: {
            solutionId: 'testId',
            solutionVersion: 'testVersion',
            sourceCodeBucket: testSourceBucket,
            sourceCodePrefix: 'testPrefix/',
        }
    });

    customResources.setupUi({
        apiEndpoint: 'https://12ab34cde5.execute-api.us-east-1.amazonaws.com/test',
        cognitoIdentityPool: 'testIdentityPool',
        cognitoUserPool: 'testUserPool',
        cognitoUserPoolClient: 'testUserPoolClient',
        consoleBucket: testConsoleBucket,
        mapName: 'testMap',
        placeIndexName: 'testPlaceIndex',
        routesBucket: testRouteBucket,
        iotPolicyName: 'testIoTPolicy'
    });

    customResources.setupDetachIotPolicyCustomResource({
        iotPolicyName: 'testIoTPolicy'
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});