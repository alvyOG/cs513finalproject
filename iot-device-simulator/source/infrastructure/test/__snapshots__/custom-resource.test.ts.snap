// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IoT Device Simulator customResourcesConstruct Test 1`] = `
Object {
  "Resources": Object {
    "TestCustomResourceConsoleConfig9233433D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ConfigFileName": "aws_config.js",
        "DestinationBucket": Object {
          "Ref": "testConsoleBucketC1925855",
        },
        "Resource": "CreateConfig",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambda4F3A7ED4",
            "Arn",
          ],
        },
        "configObj": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"aws_iot_endpoint\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "TestCustomResourceEndpointAddress55F0267E",
                  "IOT_ENDPOINT",
                ],
              },
              "\\",\\"API\\":{\\"endpoints\\":[{\\"name\\":\\"ids\\",\\"endpoint\\":\\"https://12ab34cde5.execute-api.us-east-1.amazonaws.com/test\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\"}]},\\"Auth\\":{\\"identityPoolId\\":\\"testIdentityPool\\",\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"userPoolId\\":\\"testUserPool\\",\\"userPoolWebClientId\\":\\"testUserPoolClient\\"},\\"aws_iot_policy_name\\":\\"testIoTPolicy\\",\\"aws_project_region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"geo\\":{\\"AmazonLocationService\\":{\\"region\\":\\"",
              Object {
                "Ref": "AWS::Region",
              },
              "\\",\\"maps\\":{\\"items\\":{\\"testMap\\":{\\"style\\":\\"VectorEsriNavigation\\"}},\\"default\\":\\"testMap\\"},\\"search_indices\\":{\\"items\\":[\\"testPlaceIndex\\"],\\"default\\":\\"testPlaceIndex\\"}}}}",
            ],
          ],
        },
      },
      "Type": "Custom::CopyConfigFiles",
      "UpdateReplacePolicy": "Delete",
    },
    "TestCustomResourceCopyConsoleFilesE61CD7FB": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucket": Object {
          "Ref": "testConsoleBucketC1925855",
        },
        "ManifestFile": "site-manifest.json",
        "Resource": "CopyS3Assets",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambda4F3A7ED4",
            "Arn",
          ],
        },
        "SourceBucket": "test-bucket-region",
        "SourcePrefix": "testPrefix/",
      },
      "Type": "Custom::CopyConsoleFiles",
      "UpdateReplacePolicy": "Delete",
    },
    "TestCustomResourceCopyRouteFilesB49E39F7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucket": Object {
          "Ref": "testRouteBucket64D9BEB5",
        },
        "ManifestFile": "routes-manifest.json",
        "Resource": "CopyS3Assets",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambda4F3A7ED4",
            "Arn",
          ],
        },
        "SourceBucket": "test-bucket-region",
        "SourcePrefix": "testPrefix/",
      },
      "Type": "Custom::CopyRouteFiles",
      "UpdateReplacePolicy": "Delete",
    },
    "TestCustomResourceCustomResourceLambdaIoTPolicy36A1424E": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "To connect IoT and attach IoT policy to Cognito identity cannot speficy the specific resources.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iot:DetachPrincipalPolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomResourceLambdaIoTPolicy",
        "Roles": Array [
          Object {
            "Ref": "TestCustomResourceHelperLambdaRoleA52DBB50",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestCustomResourceDetachIoTPolicyF6F108AC": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "IotPolicyName": "testIoTPolicy",
        "Resource": "DetachIoTPolicy",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambda4F3A7ED4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestCustomResourceEndpointAddress55F0267E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Resource": "DescribeIoTEndpoint",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambda4F3A7ED4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestCustomResourceHelperLambda4F3A7ED4": Object {
      "DependsOn": Array [
        "TestCustomResourceHelperLambdaRoleDefaultPolicy236FD1D8",
        "TestCustomResourceHelperLambdaRoleA52DBB50",
        "TestPolicyCC05E598",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-bucket-region",
          "S3Key": "testPrefix//custom-resource.zip",
        },
        "Description": "IoT Device Simulator custom resource function",
        "Environment": Object {
          "Variables": Object {
            "SOLUTION_ID": "testId",
            "SOLUTION_VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambdaRoleA52DBB50",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 240,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCustomResourceHelperLambdaRoleA52DBB50": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "iot:DescribeEndpoint cannot specify the resource.",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "iot:DescribeEndpoint",
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "customResourcePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestCustomResourceHelperLambdaRoleDefaultPolicy236FD1D8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::test-bucket-region",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::test-bucket-region/testPrefix//*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testConsoleBucketC1925855",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testRouteBucket64D9BEB5",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "iot:ListTargetsForPolicy",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":iot:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":policy/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCustomResourceHelperLambdaRoleDefaultPolicy236FD1D8",
        "Roles": Array [
          Object {
            "Ref": "TestCustomResourceHelperLambdaRoleA52DBB50",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestCustomResourceUUID90DFF55D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Resource": "CreateUUID",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceHelperLambda4F3A7ED4",
            "Arn",
          ],
        },
        "StackName": Object {
          "Ref": "AWS::StackName",
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "TestPolicyCC05E598": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": Array [
          Object {
            "Ref": "TestCustomResourceHelperLambdaRoleA52DBB50",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testConsoleBucketC1925855": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testRouteBucket64D9BEB5": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
